feat: reVamping the volcano, dashboard, backend, trpc, turboRepo

de6p opened the pull request:
I have completely revamped the Volcano dashboard by migrating it to a modern architecture powered by Next.js, tRPC, shadcn/ui, and Turborepo. With this upgrade, the dashboard now delivers 5 to 10 times better performance compared to the previous version. The new user interface is blazing fast, visually stunning, and highly responsive. Additionally, Turborepo significantly improves the development workflow by enabling efficient monorepo management and faster build times.


![image](https://github.com/user-attachments/assets/6c3a82b0-cd6a-4f06-a576-8d0e9020ce18)

de6p committed:
fix the create queues, and pods

Signed-off-by: Deep <de6p97@gmail.com>

jayesh9747 commented:
@de6p  i think you should create small small pr , so that it can be better to review.

de6p committed:
feat: add Dynamic fetching in the dashboard

Signed-off-by: Deep <de6p97@gmail.com>

karanBRAVO commented:
> @de6p i think you should create small small pr , so that it can be better to review.

+1

de6p committed:
feat/backend: add trpc support and improve the api response 5x-10x faster

Signed-off-by: Deep <de6p97@gmail.com>

de6p committed:
feat/frontend: revamp the volcano dashboard to newer UI system, shadcn UI, next.js

Signed-off-by: Deep <de6p97@gmail.com>

de6p committed:
feat/dashboard: add turbo repo for faster build time

Signed-off-by: Deep <de6p97@gmail.com>

JesseStutler commented:
> Hello @de6p, I have one quick question, why are we using `next.js` instead of `vite.js+express.js`?

cc @de6p

JesseStutler commented:
/gemini review

JesseStutler commented:
Also please fix the CI @de6p , and can we build the image to test now? I don't hope I meet some building errors when I want to test it

de6p committed:
feat: update ESLint configuration and dependencies, enhance Turbo build tasks, and refactor TRPC context and router logic

Signed-off-by: Deep <de6p97@gmail.com>

de6p committed:
feat: add CI workflow for ESLint, TypeScript type checking, and build process; remove outdated formatting check workflow

Signed-off-by: Deep <de6p97@gmail.com>

de6p committed:
feat: update YAML parsing in queue and pod dialogs, add js-yaml dependency, and refactor navigation icons for consistency

Signed-off-by: Deep <de6p97@gmail.com>

Monokaix commented:
Please check the gemini review comments and resolve the code conflict first.

Monokaix commented:
> @Monokaix I fixed all the comments .

Please also resolve the code conflicts: )

de6p commented:
fixed the merge conflicts.

Monokaix commented:
There some issus I currently have found:
1. The filter(Namespce, queue) func of job is lost
<img width="2236" height="688" alt="image" src="https://github.com/user-attachments/assets/bedbc9a5-e130-491f-a637-4fc4aa9e0e50" />
2. The "All" option for filter namespace is lost
<img width="2312" height="748" alt="image" src="https://github.com/user-attachments/assets/711a5578-cbca-4342-9f5b-19a6294c1f8a" />
3. The logo of dashboard keep it as volcano logo
<img width="650" height="100" alt="image" src="https://github.com/user-attachments/assets/c0cc82a0-0e96-4917-81e3-f39c568a61e3" />
4. Should also add a create job action 
<img width="2446" height="454" alt="image" src="https://github.com/user-attachments/assets/bcc5620b-7dbf-49c2-a6cd-f81e52adf85d" />
5. The dashboard page can not reflect the queue's resource, because we have removed the related reosurces in queue yaml
<img width="1216" height="1260" alt="image" src="https://github.com/user-attachments/assets/f51911e2-d56c-42fc-afa9-e355fc34a518" />
6. Some other actions are lost like delete/edit of queue/job/pod, it's already implemented in lFX term1

de6p commented:
> Why deleted test.yaml and build.yaml in workflow?

I deleted test.yaml and build.yaml because we now have a single ci.yaml workflow that runs linting, type-checking, and build steps together. This makes our CI setup simpler to maintain and ensures all checks run consistently in one place

Monokaix commented:
OK, please solve other issues.

de6p committed:
chore: update the dockerfile

Signed-off-by: Deep <de6p97@gmail.com>

de6p committed:
update the favicon to volcano logo

Signed-off-by: Deep <de6p97@gmail.com>

de6p committed:
feat: add create/delete/edit/filter support to jobs, queues, and pods

Signed-off-by: Deep <de6p97@gmail.com>

de6p committed:
fix the bar chart for queue resources

Signed-off-by: Deep <de6p97@gmail.com>

de6p committed:
feat: add podgroup dashboard support

Signed-off-by: deep <de6p97@gmail.com>
Made-with: Cursor

de6p commented:
rebased

de6p committed:
feat: implement multi-stage Docker build for Next.js app and update Kubernetes deployment configuration

Signed-off-by: Deep <de6p97@gmail.com>
Signed-off-by: deep <de6p97@gmail.com>

de6p committed:
docs: update contributing.md and fix kubernetes client config

Signed-off-by: deep <de6p97@gmail.com>
Made-with: Cursor

JesseStutler commented:
Hi @de6p the CI looks failed

de6p committed:
chore: remove outdated CI workflows and backend/frontend files

Signed-off-by: deep <de6p97@gmail.com>

de6p committed:
refactor: clean up package.json and package-lock.json by removing merge conflict markers and outdated dependencies

Signed-off-by: deep <de6p97@gmail.com>

de6p commented:
fixed ✅

JesseStutler commented on deployment/build/backend/Dockerfile:
Why the dockerfile is move to the root dir rather than here? @de6p

de6p replied:
Hey @JesseStutler  Since we moved away from separate frontend/backend apps to a single monorepo (apps/web), having per-service Dockerfiles no longer made sense. The root Dockerfile now handles the full multi-stage build for the entire monorepo — it installs dependencies via Turborepo, builds apps/web, and produces the final production image. Keeping it at the root also makes it easier for tools like Docker Buildx and CI to find it without needing to specify nested paths. Hope that clears it up!

JesseStutler commented on deployment/volcano-dashboard.yaml:
```diff
@@ -23,60 +22,40 @@ spec:
-        - image: volcanosh/vc-dashboard-frontend:latest
-          imagePullPolicy: Always
-          name: frontend
+        - name: volcano-dashboard
```
So currently we don't need frontend and backend anymore, just only on container?

de6p replied:
Hey @JesseStutler, exactly right! We've consolidated everything into a single container. The separate frontend and backend images (vc-dashboard-frontend, vc-dashboard-backend) are no longer needed — we now have one unified image (volcanosh/vc-dashboard) built from a single Dockerfile at the root. The deployment manifest will be updated to reflect just the one container. Thanks for flagging!

JesseStutler commented:
After build the image and loaded it to kind cluster and deploy it, the dashboard can't load any datas, are they are any bugs? 
<img width="2274" height="905" alt="image" src="https://github.com/user-attachments/assets/66fe6ea5-9502-49e9-b9fa-461076db3825" />
It always got stuck here @de6p

de6p commented:
1. Build dashboard image locally:
```bash
docker build -t dashboard-app:dev .
```

2. Load image into your kind cluster:
```bash
kind load docker-image dashboard-app:dev --name kind
```

3. Deploy dashboard resources:
```bash
kubectl create ns volcano-system --dry-run=client -o yaml | kubectl apply -f -
kubectl apply -f deployment/volcano-dashboard.yaml
```

4. Point deployment to your local image:
```bash
kubectl -n volcano-system set image deploy/volcano-dashboard volcano-dashboard=dashboard-app:dev
kubectl -n volcano-system rollout status deploy/volcano-dashboard
```

5. Install Volcano CRDs/controllers (the missing piece):
```bash
kubectl apply -f https://raw.githubusercontent.com/volcano-sh/volcano/master/installer/volcano-development.yaml
```

6. Verify Volcano APIs now exist:
```bash
kubectl api-resources | grep -E "batch.volcano.sh|scheduling.volcano.sh"
kubectl get jobs.batch.volcano.sh -A
kubectl get queues.scheduling.volcano.sh
kubectl get podgroups.scheduling.volcano.sh -A
```

7. Access dashboard:
```bash
kubectl -n volcano-system port-forward svc/volcano-dashboard 8080:80
```
Open `http://localhost:8080`.

---

JesseStutler commented on deployment/volcano-dashboard.yaml:
```diff
@@ -23,60 +22,40 @@ spec:
           ports:
-            - containerPort: 8080
-              name: frontend
+            - containerPort: 3000
```
Why does the port needs be migrated? @de6p

de6p replied:
No K8s requirement—we only need two containerPorts if the pod actually listens on two ports. Our image is a single Next server on 3000, so we’ll drop the 8080 entry and keep everything on 3000 to match the container.

de6p replied:
if you wanted, we could change it as well.

JesseStutler replied:
I think we should keep the previous behavior, change the port will be breaking changes when users have already used the old port

de6p replied:
ok, i moved back to 8080

JesseStutler commented on deployment/volcano-dashboard.yaml:
```diff
@@ -23,60 +22,40 @@ spec:
-          imagePullPolicy: Always
-          name: frontend
+        - name: volcano-dashboard
+          image: volcanosh/volcano-dashboard:latest
```
Just a small requirement besides this pr, could you also add a bump script just like volcano main repo, when we need to release a new tag of dashboard, can the tags here be replaced with one click? This can be followed up in a later PR.

de6p replied:
added ✅

JesseStutler replied:
Where did you add? I didn't see it

de6p replied:
it is under /hack/bump-dashboard-version.sh

de6p committed:
feat: enhance CreateQueueDialog with YAML support and resource management features

- Added functionality to parse and build queue manifests from form inputs and YAML.
- Introduced resource pair handling for CPU and memory specifications.
- Implemented form reset and state management for queue creation.
- Added a new script to bump dashboard version tags in YAML files.

Signed-off-by: deep <de6p97@gmail.com>

de6p committed:
remove: dialog box descriptions

Signed-off-by: deep <de6p97@gmail.com>

de6p committed:
remove: unused variable

Signed-off-by: deep <de6p97@gmail.com>

JesseStutler commented on apps/web/src/app/(dashboard)/jobs/page.tsx:
Why are there parentheses around (dashboard) in the path? @de6p

de6p replied:
`(dashboard)` is a Next.js route group — it helps organize related pages/layouts while keeping the URL clean. So /jobs stays /jobs, not /dashboard/jobs.

JesseStutler commented on deployment/Dockerfile:
Why the Dockerfile is under root dir but volcano-dashboard.yaml is still in the deployment dir? I prefer Dockerfile should still in the deployment dir

de6p replied:
moved to deployment dir and updated the docs as well

JesseStutler commented on deployment/volcano-dashboard.yaml:
```diff
@@ -23,60 +22,40 @@ spec:
-          name: frontend
+        - name: volcano-dashboard
+          image: volcanosh/volcano-dashboard:latest
+          imagePullPolicy: Never
```
Should be IfNotPresent

de6p replied:
ok

de6p committed:
feat: implement pagination component for dashboard tables, and fix the API

Signed-off-by: Deep <de6p97@gmail.com>

de6p committed:
chore: update Dockerfile path and adjust deployment configurations

Signed-off-by: Deep <de6p97@gmail.com>

JesseStutler commented on deployment/volcano-dashboard.yaml:
```diff
@@ -23,60 +22,40 @@ spec:
-          imagePullPolicy: Always
-          name: frontend
+        - name: volcano-dashboard
+          image: volcanosh/volcano-dashboard:v1.2.0
```
@de6p I think here should be latest, Firstly, we don't have a v1.2.0 image at all, only v0.2.0. Secondly, we should keep it as the latest image here because this is code being merged into the main branch, not release branch. Using an older image might result in the YAML being modified, but the image itself remaining old and unusable.

de6p replied:
fixed

de6p committed:
chore: update volcano-dashboard image tag to latest

Signed-off-by: deep <de6p97@gmail.com>
